pkg/controllers/management/auth/project_cluster/cluster_handler.go

```diff
@@ -0,0 +1,417 @@
+			return obj, err
+		}
+
+		annotation := map[string]string{}
+
+		creatorID := metaAccessor.GetAnnotations()[CreatorIDAnn]
+		if creatorID != "" {
+			annotation[CreatorIDAnn] = creatorID
+		}
+
+		if name == project.System {
+			latestSystemVersion, err := systemimage.GetSystemImageVersion()
+			if err != nil {
+				return obj, err
+			}
+			annotation[project.SystemImageVersionAnn] = latestSystemVersion
+		}
+
+		project := &apisv3.Project{
+			ObjectMeta: v1.ObjectMeta{
+				GenerateName: "p-",
+				Annotations:  annotation,
+				Labels:       labels,
+			},
+			Spec: apisv3.ProjectSpec{
+				DisplayName: name,
+				Description: fmt.Sprintf("%s project created for the cluster", name),
+				ClusterName: metaAccessor.GetName(),
+			},
+		}
+		updated, err := l.addRTAnnotation(project, "project")
+		if err != nil {
+			return obj, err
+		}
+		project = updated.(*apisv3.Project)
+		logrus.Infof("[%v] Creating %s project for cluster %v", ClusterCreateController, name, metaAccessor.GetName())
+		if _, err = l.mgr.Management.Projects(metaAccessor.GetName()).Create(project); err != nil {
+			return obj, err
+		}
+		return obj, nil
+	})
+}
+
+// deleteSystemProject deletes the system project(s) for a cluster in preparation for deleting the cluster namespace.
+// Normally, the webhook prevents deleting the system project, so Rancher needs to use the sudo user to force it.
+// Otherwise, the deleted namespace will be stuck terminating because it cannot garbage collect the project.
+func (l *clusterLifecycle) deleteSystemProject(cluster *apisv3.Cluster, controller string) error {
+	bypassClient, err := l.projects.WithImpersonation(controllers.WebhookImpersonation())
+	if err != nil {
+		return fmt.Errorf("[%s] failed to create impersonation client: %w", controller, err)
+	}
+	projects, err := l.projectLister.List(cluster.Name, systemProjectLabels.AsSelector())
+	if err != nil {
+		return fmt.Errorf("[%s] failed to list projects: %w", controller, err)
+	}
+	var deleteError error
+	for _, p := range projects {
+		logrus.Infof("[%s] Deleting project %s", controller, p.Name)
+		err = bypassClient.Delete(p.Namespace, p.Name, nil)
+		if err != nil {
+			deleteError = errors.Join(deleteError, fmt.Errorf("[%s] failed to delete project '%s/%s': %w", controller, p.Namespace, p.Name, err))
+		}
+	}
+	return deleteError
+}
+
+func (l *clusterLifecycle) addRTAnnotation(obj runtime.Object, context string) (runtime.Object, error) {
+	meta, err := meta.Accessor(obj)
+	if err != nil {
+		return obj, err
+	}
+
+	// If the annotation is already there move along
+	if _, ok := meta.GetAnnotations()[roleTemplatesRequired]; ok {
+		return obj, nil
+	}
+
+	rt, err := l.roleTemplateLister.List("", labels.NewSelector())
+	if err != nil {
+		return obj, err
+	}
+
+	annoMap := make(map[string][]string)
+
+	var restrictedAdmin bool
+	if settings.RestrictedDefaultAdmin.Get() == "true" {
+		restrictedAdmin = true
+	}
+
+	annoMap["created"] = []string{}
```

Comment:
Looking into this

Reply:
I'm not sure if I totally follow. Is your preference something like this https://go.dev/play/p/WkZUQmRGYTO? I'm not sure if it really simplifies the code much

Reply:
You can simplify that further, but it would leave the `created` as a `nil` slice.

So, maybe just a comment to say that `created` isn't used, but it's required in the data?

(One arguable benefit of using a stuct is that it can be unmarshaled to a value of the struct type too, rather than the current keys), see "Connascence of Meaning" :-)

Reply:
I opted to just add the comment for now